core/array.cpp

```diff
@@ -108,12 +112,34 @@ uint32_t Array::hash() const {
 	}
 	return h;
 }
-void Array::operator=(const Array &p_array) {
 
-	_ref(p_array);
+void Array::_assign(const Array &p_array) {
+
+	if (_p->typed.type != Variant::OBJECT && _p->typed.type == p_array._p->typed.type) {
+		//same type or untyped, just reference, shuold be fine
+		_ref(p_array);
+	} else if (_p->typed.type == Variant::NIL) { //from typed to untyped, must copy, but this is cheap anyway
+		_p->array = p_array._p->array;
+	} else if (p_array._p->typed.type == Variant::NIL) { //from untyped to typed, must try to check if they are all valid
+		for (int i = 0; i < p_array._p->array.size(); i++) {
+			if (!_p->typed.validate(p_array._p->array[i], "assign")) {
+				return;
+			}
+		}
+		_p->array = p_array._p->array; //then just copy, which is cheap anyway
```

Comment:
I'm worried about two things here:
1. This looks slow. Validating every single element on assignment and then copying.
2. It's inconsistent with untyped arrays, which are referenced instead of copied.

Reply:
Ah, wait. That makes more sense. Let's see. I think we shouldn't allow assigning from untyped to typed. If such a thing is desired, it should be done explicitly with clone/duplicate.
I still don't like the validation penalty, but I can't think of a better solution and I think it's a very uncommon conversion any way.

Reply:
I think untyped to typed is fine, even with the penalty. In most cases, users will just pass [1,2,3] anyway to a funcion that takes a typed array and if the types are wrong, it will fail anyway. 

Remember that this is mostly meant for GDScript, which you should be able to use untyped as much as you want. In Mono you can expose those types properly now, so C# side of things will not be able to send or receive untyped anyway.

Reply:
Yes, this is not something that would affect C#.
Still, from a GDScript point of view, what really worries me is the inconsistency between assigning by ref and by copy here. I can see people doing assignments where they expect a copy but get a ref instead (or viceversa) and get all confused about it.

Reply:
I suppose we can detect this in the parser and add a warning, should probably be easy for @vnen. If you do it you get a warning, and if you really meant to do it, you disable the warning.